joshua/joshua_model.py

```diff
@@ -563,25 +569,91 @@ def set_versionstamped_key(tr, prefix, suffix, value):
     tr.add(dir_all_ensembles[ensemble_id]['count'][counter], ONE)
 
 
-def _add(tr, ensemble_id, counter, value):
+def _add(tr : fdb.Transaction, ensemble_id : str, counter : str, value : int) -> None:
     byte_val = struct.pack("<Q", value)
     tr.add(dir_all_ensembles[ensemble_id]['count'][counter], byte_val)
 
 
-def _get_snap_counter(tr, ensemble_id, counter):
+def _get_snap_counter(tr : fdb.Transaction, ensemble_id : str, counter : str) -> int:
     value = tr.snapshot.get(dir_all_ensembles[ensemble_id]['count'][counter])
     if value == None:
         return 0
     else:
         return struct.unpack("<Q", b'' + value)[0]
 
 
+def _get_snap_max_runs(tr : fdb.Transaction, ensemble_id : str) -> int:
+    """Read max_runs for this ensemble at snapshot isolation. Returns a default of 0 if max_runs is not set"""
+    result = tr.snapshot.get(dir_all_ensembles[ensemble_id]['properties']['max_runs'])
+    if result == None:
+        return 0
+    value, = fdb.tuple.unpack(result)
+    return value
+
+def _get_snap_timeout(tr : fdb.Transaction, ensemble_id : str) -> int:
+    """Read timeout for this ensemble at snapshot isolation. Returns a default of 5400 if timeout is not set"""
+    result = tr.snapshot.get(dir_all_ensembles[ensemble_id]['properties']['timeout'])
+    if result == None:
+        return 0
+    value, = fdb.tuple.unpack(result)
+    return value
+
+class EnsembleProgressTracker:
+    def __init__(self):
+        # map from ensemble_id to last known ended count and the most recent time we
+        # observed the count change. Ordered by last updated.
+        self._last_ensemble_progress : OrderedDict[str, Tuple[int, float]] = OrderedDict()
+
+    def try_start(self, ensemble_id : str, tr : fdb.Transaction) -> bool:
+        """
+        Return True if this agent should start a run for this ensemble. The
+        policy tries not to overshoot max_runs by too much, but also accounts
+        for the possibility that agents might die.
+        """
+        started = _get_snap_counter(tr, ensemble_id, "started")
+        ended = _get_snap_counter(tr, ensemble_id, "ended")
+        max_runs = _get_snap_max_runs(tr, ensemble_id)
+        timeout = _get_snap_timeout(tr, ensemble_id)
+        # max_runs == 0 means run forever
+        if max_runs > 0 and started >= max_runs:
+            if ensemble_id in self._last_ensemble_progress:
+                (last_ended, last_time) = self._last_ensemble_progress[ensemble_id]
+                if ended > last_ended:
+                    self._last_ensemble_progress[ensemble_id] = ended, time.time()
+                    self._last_ensemble_progress.move_to_end(ensemble_id)
+                    return False
+                elif time.time() - last_time < timeout + 10: # 10 second buffer to give the other agent a chance to end their run
+                    # Don't start it yet - maybe the other agent hasn't timed out yet
+                    time.sleep(1) # Avoid polling too frequently
```

Comment:
If the only ensemble left is one where each test runs for 15 minutes, and has started a sufficient number of jobs, then this would still leave all joshua agents logging "<job stopped or enough runs started>" every 1s.

Is there any form of this check that could be added to list_and_watch_active_ensembles so that the agent doesn't call run_ensemble on ensembles that we don't want to be running?  Or maybe the call to this should function be lifted out of run_ensemble and done where one calls run_ensemble instead?

Reply:
I made an attempt to lift the check into the main run loop in joshua_agent. PTAL @sfc-gh-almiller